storage/src/sparse_merkle/internal_node.cpp

```diff
@@ -0,0 +1,171 @@
+    auto& parent = children_.at(parent_idx);
+    if (!parent) {
+      // We are adding the first child below a non-existent parent. Create the
+      // parent first.
+      children_[parent_idx] = InternalChild{Hash(), version};
+      parent = children_.at(parent_idx);
+    }
+    // A parent is *always* an InternalChild
+    auto& internal_child = std::get<InternalChild>(parent.value());
+    internal_child.hash = parent_hash;
+    internal_child.version = version;
+    if (parent_idx == 0) {
+      // We have reached the root of this BatchedInternalNode
+      return;
+    }
+    index = parent_idx;
+  }
+}
+
+BatchedInternalNode::InsertResult BatchedInternalNode::insert(LeafChild& child, size_t depth) {
+  // The index into the children_ array
+  size_t index = 0;
+  Nibble child_key = child.key.hash().get_nibble(depth);
+
+  for (size_t i = Nibble::SIZE_IN_BITS - 1; i >= 0 && i != SIZE_MAX; i--) {
+    if (child_key.get_bit(i)) {
+      index = right_child_index(index);
+    } else {
+      index = left_child_index(index);
+    }
+    if (is_internal(index)) {
+      continue;
+    }
+
+    // First try to insert into an empty child, then try to overwrite the child,
+    // and lastly split the node.
+    if (auto rv = insert_into_empty_child(index, child)) {
+      return rv.value();
+    }
+    if (auto rv = overwrite(index, child)) {
+      return rv.value();
+    }
+    return split_node(index, child, depth);
+  }
+
+  // We have reached the leaf of this BatchedInternalNode and it points to another
+  // BatchedInternalNode. Return the Version so the caller can construct an
+  // InternalNodeKey and retrieve the next batched node in the tree.
+  auto& stored_child = children_[index];
+  return BatchedInternalNode::InsertIntoExistingNode{std::get<InternalChild>(stored_child.value()).version};
+}
+
+std::optional<BatchedInternalNode::InsertResult> BatchedInternalNode::insert_into_empty_child(size_t index,
+                                                                                              LeafChild& child) {
+  if (is_empty(index)) {
+    auto version = child.key.version();
+    children_[index] = child;
+    update_hashes(index, version);
+    return BatchedInternalNode::InsertComplete{};
+  }
+  return std::nullopt;
+}
+
+std::optional<BatchedInternalNode::InsertResult> BatchedInternalNode::overwrite(size_t index, LeafChild& child) {
+  auto& stored_child = std::get<LeafChild>(children_.at(index).value());
+  if (stored_child.key.hash() == child.key.hash()) {
+    auto result = BatchedInternalNode::InsertComplete{stored_child.key};
+    stored_child = child;
+    update_hashes(index, child.key.version());
+    return result;
+  }
+  return std::nullopt;
+}
+
+BatchedInternalNode::InsertResult BatchedInternalNode::split_node(size_t index, LeafChild& child, size_t depth) {
+  auto version = child.key.version();
+  auto stored_leaf_child = std::get<LeafChild>(children_.at(index).value());
```

Comment:
in overwrite() function you gave a hint: auto**&** stored_child. It was better than here.

Reply:
Two different things are happening in these functions. In overwrite, we need a reference so we can **overwrite** the location in `children_`. We don't want a reference here though, we want a copy. If we take a reference, then when we put a new `InternalChild` into `children_`, the reference will contain the new value, and we will have lost the old value that we need for use later in the function. Since the node is being split we need to save the old value so we can insert it later or return it to the caller to insert in a new `BatchedInternalNode`.